src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -275,20 +275,29 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
            string filterText,
            IList<MatchResult> builder)
         {
+            using var _1 = ArrayBuilder<CompletionItem>.GetInstance(matchResults.Count, out var itemBuilder);
+            using var _2 = PooledDictionary<CompletionItem, MatchResult>.GetInstance(out var map);
```

Comment:
All internal implementations of CompletionService should override this for better porformance. I will fix TS once this is inserted.

Reply:
Is the public overload marked as [Obsolete]?

Reply:
It's not, since we don't have an alternative available publicly yet.

Reply:
that seems problematic to me.  this is intended to be an open api.  if we lock performance behind only things we (and our first party partners) can acquire, that's really unfortunate :(

Reply:
May be we can figure something out about this, but tight now we can't expose it because types like PatternMatch are not public, and MatchResult being a value type doesn't help either

Reply:
why does filtering items need pattern match results?

Reply:
Actually, forget what I said earlier. The public `FilterItems` API is fine. `CompletionService` cannot be implemented externally and anyone who host Roslyn and uses the public overload of `FilterItems` would have no negative perf imapct.

The only problem here is when CompletionService is hosted in VS, our AsyncCompletion layer calls the internal `FilterItems` overload to ensure best performance for large amount of CompletionItems. But TS implements `CompletionService` via ExternalAccess and they override the public `FilterItems` method. Because of this, TS has to go through the default implementation of internal `FilterItems` and pay the overhead of calling into the public `FilterItems` method. (C#/VB CompletionService override the default thus avoided this)

This PR is to provide TS a way to override the internal API instead of the public one.